File: src/main/java/ch/obermuhlner/astro/app/Astrophotography.java
package ch.obermuhlner.astro.app;

import ch.obermuhlner.astro.app.CommandParser.Command;
import ch.obermuhlner.astro.app.CommandParser.Option;
import ch.obermuhlner.astro.gradient.Point;
import ch.obermuhlner.astro.gradient.filter.Filter;
import ch.obermuhlner.astro.gradient.filter.GaussianBlurFilter;
import ch.obermuhlner.astro.gradient.filter.GradientInterpolationFilter;
import ch.obermuhlner.astro.gradient.filter.HorizontalMedianFilter;
import ch.obermuhlner.astro.gradient.filter.MedianFilter;
import ch.obermuhlner.astro.gradient.filter.PseudoMedianFilter;
import ch.obermuhlner.astro.gradient.filter.VerticalMedianFilter;
import ch.obermuhlner.astro.gradient.operation.ImageOperation;
import ch.obermuhlner.astro.gradient.operation.SubtractLinearImageOperation;
import ch.obermuhlner.astro.image.color.ColorModel;
import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.ImageCreator;
import ch.obermuhlner.astro.image.ImageQuality;
import ch.obermuhlner.astro.image.ImageReader;
import ch.obermuhlner.astro.image.ImageWriter;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.regex.Pattern;

public class Astrophotography {

  public static void mainTest(String[] args) throws IOException, ScriptException {
    runTest();
  }

  public static void main(String[] args) throws IOException, ScriptException {
    String outputFilePrefix = "output_";
    List<File> inputFiles = new ArrayList<>();
    List<File> outputFiles = new ArrayList<>();

    AtomicReference<Filter> gradientFilter = new AtomicReference<>(new GaussianBlurFilter(100, ColorModel.RGB));
    AtomicReference<ImageOperation> subtractOperation = new AtomicReference<>(new SubtractLinearImageOperation());
    List<Point> gradientPoints = new ArrayList<>();
    AtomicInteger sampleRadius = new AtomicInteger(3);

    CommandParser commandParser = new CommandParser();

    commandParser.add(
        new Command("script", 1),
        (commandArguments, optionsWithArguments) -> {
          String[] script = loadScript(new File(commandArguments.arguments.get(0)));
          commandParser.parse(script);
        }
    );
    commandParser.add(
        new Command("input", 1),
        (commandArguments, optionsWithArguments) -> {
          inputFiles.add(new File(commandArguments.arguments.get(0)));
        }
    );
    commandParser.add(
        new Command("batch", 0,
            new Option("file", 1)),
        (commandArguments, optionsWithArguments) -> {
          optionsWithArguments.handleOption("file", arguments -> {
            inputFiles.add(new File(arguments.get(0)));
          });
        }
    );
    commandParser.add(
        new Command("gradient", 0,
            new Option("point", 1),
            new Option("sampleradius", 1),
            new Option("interpolationpower", 1)
        ),
        (commandArguments, optionsWithArguments) -> {
          GradientInterpolationFilter gradientInterpolationFilter = new GradientInterpolationFilter();
          optionsWithArguments.handleOption("point", arguments -> {
            gradientPoints.add(toPoint(arguments.get(0)));
          });
          optionsWithArguments.handleOption("interpolationpower", arguments -> {
            gradientInterpolationFilter.setInterpolationPower(Double.parseDouble(arguments.get(0)));
          });
          optionsWithArguments.handleOption("sampleradius", arguments -> {
            sampleRadius.set(Integer.parseInt(arguments.get(0)));
          });

          gradientFilter.set(gradientInterpolationFilter);
        }
    );
    commandParser.add(
        new Command("output", 1),
        (commandArguments, optionsWithArguments) -> {
          outputFiles.add(new File(commandArguments.arguments.get(0)));
        }
    );

    commandParser.parse(new String[] {
        "input", "images/Autosave001.tif",
//        "input", "images/inputs/Autosave001_small_compress0.png",
//        "median-blur", "10",
//        "gaussian-blur", "50",
//        "gradient", "--point", "100,100", "--point", "-100,-100",
        "output", "images/Test.png"

    });
    //commandParser.parse(args);

    for (int f = 0; f < inputFiles.size(); f++) {
      File inputFile = inputFiles.get(f);
      File outputFile;
      if (f < outputFiles.size()) {
        outputFile = outputFiles.get(f);
      } else {
        outputFile = new File(inputFile.getParent(), outputFilePrefix + inputFile.getName());
      }

      System.out.println("Load " + inputFile);

      DoubleImage inputImage = ImageReader.read(inputFile, ImageQuality.High);

      if (gradientFilter.get() instanceof GradientInterpolationFilter) {
        GradientInterpolationFilter gradientInterpolationFilter = (GradientInterpolationFilter) gradientFilter.get();
        gradientInterpolationFilter.setFixPoints(correctFixPoints(gradientPoints, inputImage), inputImage, sampleRadius.get());
      }

      System.out.println("Create gradient " + gradientFilter.get());
      DoubleImage gradientImage = gradientFilter.get().filter(inputImage);

      System.out.println("Subtract gradient " + subtractOperation.get());
      DoubleImage outputImage = ImageCreator.create(inputImage.getWidth(), inputImage.getHeight(), ImageQuality.High);
      subtractOperation.get().operation(inputImage, gradientImage, outputImage);

      System.out.println("Save " + outputFile);
      ImageWriter.write(outputImage, outputFile);

      System.out.println("Finished " + inputFile + " -> " + outputFile);
    }
  }

  private static void autoSetFixPoints(GradientInterpolationFilter gradientInterpolationFilter, DoubleImage image) {
    int sampleWidth = image.getWidth() / 5;
    int sampleHeight = image.getHeight() / 5;

    int x1 = image.getWidth() / 2;
    int y1 = sampleHeight / 2;
    double[] color1 = image.medianPixel(x1, y1, sampleWidth, sampleHeight, ColorModel.RGB, null);
    System.out.println("Auto median pixel1: " + Arrays.toString(color1));

    int x2 = sampleWidth / 2;
    int y2 = image.getHeight() - sampleHeight / 2;
    double[] color2 = image.medianPixel(x2, y2, sampleWidth, sampleHeight, ColorModel.RGB, null);
    System.out.println("Auto median pixel2: " + Arrays.toString(color2));

    int x3 = image.getWidth() - sampleWidth / 2;
    int y3 = image.getHeight() - sampleHeight / 2;
    double[] color3 = image.medianPixel(x3, y3, sampleWidth, sampleHeight, ColorModel.RGB, null);
    System.out.println("Auto median pixel3: " + Arrays.toString(color3));

    gradientInterpolationFilter.setFixPoints(
        Arrays.asList(new Point(x1, y1), new Point(x2, y2), new Point(x3, y3)),
        Arrays.asList(color1, color2, color3));
  }

  private static String[] loadScript(File file) {
    try {
      String script = Files.readString(file.toPath());
      return script.split("\\s+");
    }
    catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private static List<Point> correctFixPoints(List<Point> fixPoints, DoubleImage inputImage) {
    List<Point> result = new ArrayList<>();

    for (Point fixPoint : fixPoints) {
      if (fixPoint.x < 0) {
        fixPoint = new Point(inputImage.getWidth() + fixPoint.x, fixPoint.y);
      }
      if (fixPoint.y < 0) {
        fixPoint = new Point(fixPoint.x, inputImage.getHeight() + fixPoint.y);
      }

      result.add(fixPoint);
    }

    return result;
  }

  private static Point toPoint(String string) {
    String[] split = string.split(Pattern.quote(","));
    return new Point(Integer.parseInt(split[0]), Integer.parseInt(split[1]));
  }

  private static void runTest() throws IOException {
    var input = loadImage("images/Autosave001.tif");
    var gradient = gaussianBlur(input, 20);
    saveImage(gradient, "images/TestGradient.png");
  }

  private static void runScript() throws ScriptException {
    ScriptEngineManager manager = new ScriptEngineManager();
    ScriptEngine engine = manager.getEngineByName("jshell");

    String script = "" +
        "import static ch.obermuhlner.astro.app.Astrophotography.*;" +
        "var input = loadImage(\"images/Autosave001.tif\");" +
        "var gradient = gaussianBlur(input, 200);" +
        "var output = subtractLinear(input, gradient);" +
        "saveImage(output, \"images/Test.png\");";

    engine.eval(script);
  }

  public static DoubleImage loadImage(String filename) throws IOException {
    return ImageReader.read(new File(filename), ImageQuality.High);
  }
  public static void saveImage(DoubleImage image, String filename) throws IOException {
    ImageWriter.write(image, new File(filename));
  }
  public static DoubleImage gaussianBlur(DoubleImage image, int radius) {
    return new GaussianBlurFilter(radius, ColorModel.RGB).filter(image);
  }
  public static DoubleImage medianBlur(DoubleImage image, int radius) {
    return new MedianFilter(radius, ColorModel.RGB).filter(image);
  }
  public static DoubleImage horizontalMedianFilter(DoubleImage image, int radius) {
    return new HorizontalMedianFilter(radius, ColorModel.RGB).filter(image);
  }
  public static DoubleImage verticalMedianFilter(DoubleImage image, int radius) {
    return new VerticalMedianFilter(radius, ColorModel.RGB).filter(image);
  }
  public static DoubleImage pseudoMedianFilter(DoubleImage image, int radius) {
    return new PseudoMedianFilter(radius, ColorModel.RGB).filter(image);
  }
  public static DoubleImage gradient(DoubleImage image) {
    GradientInterpolationFilter gradientInterpolationFilter = new GradientInterpolationFilter();
    autoSetFixPoints(gradientInterpolationFilter, image);
    return gradientInterpolationFilter.filter(image);
  }
  public static DoubleImage subtractLinear(DoubleImage image1, DoubleImage image2) {
    return new SubtractLinearImageOperation().operation(image1, image2);
  }

}

File: src/main/java/ch/obermuhlner/astro/app/Astrophotography.kt
package ch.obermuhlner.astro.app

import ch.obermuhlner.astro.gradient.Point
import ch.obermuhlner.astro.gradient.filter.*
import ch.obermuhlner.astro.gradient.operation.SubtractImageOperation
import ch.obermuhlner.astro.gradient.operation.SubtractLinearImageOperation
import ch.obermuhlner.astro.gradient.operation.SubtractSplineImageOperation
import ch.obermuhlner.astro.image.*
import ch.obermuhlner.astro.image.color.ColorModel
import com.xenomachina.argparser.ArgParser
import com.xenomachina.argparser.default
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.util.*
import java.util.regex.Pattern
import javax.script.ScriptEngineManager
import javax.script.ScriptException

enum class GlowStrategy {
    Gradient,
    Blur,
    SingleColor
}

enum class SingleColorStrategy {
    Median,
    Average,
    Darkest
}

enum class SubtractStrategy {
    Subtract,
    SubtractLinear,
    SubtractSpline
}

object Astrophotography {
    @Throws(IOException::class)
    @JvmStatic
    fun mainTEST(args: Array<String>) {
        runTest()
    }

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val parser = ArgParser(args)
//        val parser = ArgParser(arrayOf(
//                "-v",
//                "--blur",
//                "images/Autosave001.tif"
//        ))

        val verbose by parser.flagging(
                "-v", "--verbose",
                help = "enable verbose mode")

        val glowStrategy by parser.mapping(
                "--gradient" to GlowStrategy.Gradient,
                "--blur" to GlowStrategy.Blur,
                "--single-color" to GlowStrategy.SingleColor,
                help = "glow removal strategy").default(GlowStrategy.Gradient)

        val points by parser.adding(
                "-p", "--point",
                help = "fix point to determine glow gradient") { toPoint(this) }

        val sampleRadius by parser.storing(
                "--sample-radius",
                help = "sample radius") { toInt() }.default(3)

        val interpolationPower by parser.storing(
                "--interpolation-power",
                help = "interpolation power") { toDouble() }.default(3.0)

        val despeckleRadius by parser.storing(
                "--despeckle-radius",
                help = "despeckle radius") { toInt() }.default(5)

        val blurRadius by parser.storing(
                "--blur-radius",
                help = "blur radius") { toInt() }.default(100)

        val singleColorStrategy by parser.mapping(
                "--median-color" to SingleColorStrategy.Median,
                "--average-color" to SingleColorStrategy.Average,
                "--darkest-color" to SingleColorStrategy.Darkest,
                help = "").default(SingleColorStrategy.Median)

        val subtractStrategy by parser.mapping(
                "--subtract" to SubtractStrategy.Subtract,
                "--subtract-linear" to SubtractStrategy.SubtractLinear,
                "--subtract-spline" to SubtractStrategy.SubtractSpline,
                help = "").default(SubtractStrategy.SubtractLinear)

        val splineFactor by parser.storing(
                "--spline-factor",
                help = "spline factor") { toDouble() }.default(0.01)

        val sources by parser.positionalList(
                help = "source filename")

        if (verbose) {
            println("glowStrategy: $glowStrategy")
            println("points: $points")
            println("sampleRadius: $sampleRadius")
            println("interpolationPower: $interpolationPower")
            println("despeckleRadius: $despeckleRadius")
            println("blurRadius: $blurRadius")
            println("singleColorStrategy: $singleColorStrategy")
            println("sources: $sources")
        }

        for (source in sources) {
            println("Loading $source")
            val sourceFile = File(source)
            val inputImage = ImageReader.read(sourceFile, ImageQuality.High)

            println("Processing $source with $glowStrategy")
            val gradientImage = when (glowStrategy) {
                GlowStrategy.SingleColor -> {
                    val color = pickColor(inputImage, singleColorStrategy)
                    println("Picking $singleColorStrategy: ${color.contentToString()}")
                    val colorImage = ArrayDoubleImage(inputImage.width, inputImage.height, ColorModel.RGB)
                    colorImage.setPixels(color)
                    colorImage
                }
                GlowStrategy.Blur -> {
                    println("Despeckling $despeckleRadius and blurring $blurRadius pixels")
                    val despeckledImage = PseudoMedianFilter(despeckleRadius, ColorModel.RGB).filter(inputImage)
                    GaussianBlurFilter(blurRadius, ColorModel.RGB).filter(despeckledImage)
                }
                GlowStrategy.Gradient -> {
                    val gradientFilter = GradientInterpolationFilter(interpolationPower)
                    if (points.isEmpty()) {
                        println("Interpolating gradient between automatically determined points")
                        autoSetFixPoints(gradientFilter, inputImage)
                    } else {
                        val fixPoints = correctFixPoints(points, inputImage)
                        println("Interpolating gradient between $fixPoints")
                        gradientFilter.setFixPoints(fixPoints, inputImage, sampleRadius)
                    }
                    gradientFilter.filter(inputImage)
                }
            }

            val gradientFile = File(sourceFile.parent, "gradient_" + sourceFile.name)
            println("Saving $gradientFile")
            ImageWriter.write(gradientImage, gradientFile)

            println("Subtracting calculated glow from input image")
            val subtractOperation = when (subtractStrategy) {
                SubtractStrategy.Subtract -> {
                    SubtractImageOperation()
                }
                SubtractStrategy.SubtractLinear -> {
                    SubtractLinearImageOperation()
                }
                SubtractStrategy.SubtractSpline -> {
                    SubtractSplineImageOperation(splineFactor)
                }
            }
            val outputImage = ImageCreator.create(inputImage.width, inputImage.height, ImageQuality.High)
            subtractOperation.operation(inputImage, gradientImage, outputImage)

            val outputFile = File(sourceFile.parent, "output_" + sourceFile.name)
            println("Saving $outputFile")
            ImageWriter.write(outputImage, outputFile)
        }
    }

    private fun pickColor(image: DoubleImage, singleColorStrategy: SingleColorStrategy): DoubleArray {
        return when (singleColorStrategy) {
            SingleColorStrategy.Median -> {
                image.medianPixel()
            }
            SingleColorStrategy.Average -> {
                image.averagePixel()
            }
            SingleColorStrategy.Darkest -> {
                image.darkestPixel()
            }
        }
    }

    private fun autoSetFixPoints(gradientInterpolationFilter: GradientInterpolationFilter, image: DoubleImage) {
        val sampleWidth = image.width / 5
        val sampleHeight = image.height / 5
        val x1 = image.width / 2
        val y1 = sampleHeight / 2
        val color1 = image.medianPixel(x1, y1, sampleWidth, sampleHeight)
        println("Auto median pixel1: " + color1.contentToString())
        val x2 = sampleWidth / 2
        val y2 = image.height - sampleHeight / 2
        val color2 = image.medianPixel(x2, y2, sampleWidth, sampleHeight)
        println("Auto median pixel2: " + color2.contentToString())
        val x3 = image.width - sampleWidth / 2
        val y3 = image.height - sampleHeight / 2
        val color3 = image.medianPixel(x3, y3, sampleWidth, sampleHeight)
        println("Auto median pixel3: " + color3.contentToString())
        gradientInterpolationFilter.setFixPoints(
                listOf(Point(x1, y1), Point(x2, y2), Point(x3, y3)),
                listOf(color1, color2, color3))
    }

    private fun correctFixPoints(fixPoints: List<Point>, inputImage: DoubleImage): List<Point> {
        val result: MutableList<Point> = ArrayList()
        for (fixPoint in fixPoints) {
            var fp = fixPoint
            if (fp.x < 0) {
                fp = Point(inputImage.width + fp.x, fp.y)
            }
            if (fp.y < 0) {
                fp = Point(fixPoint.x, inputImage.height + fp.y)
            }
            result.add(fp)
        }
        return result
    }

    private fun toPoint(string: String): Point {
        val split = string.split(Pattern.quote(",").toRegex()).toTypedArray()
        return Point(split[0].toInt(), split[1].toInt())
    }

    @Throws(IOException::class)
    private fun runTest() {
        val input = loadImage("images/Autosave001.tif")
        val nostars1 = pseudoMedianFilter(input, 20)
        saveImage(nostars1, "images/TestNoStars1.png")
        val nostars2 = gaussianBlur(nostars1, 2)
        saveImage(nostars2, "images/TestNoStars2.png")
        val stars1 = subtractLinear(input, nostars1)
        saveImage(stars1, "images/TestStars1.png")
        val stars2 = subtractLinear(input, nostars2)
        saveImage(stars2, "images/TestStars2.png")
    }

    @Throws(ScriptException::class)
    private fun runScript() {
        val manager = ScriptEngineManager()
        val engine = manager.getEngineByName("jshell")
        val script = "" +
                "import static ch.obermuhlner.astro.app.Astrophotography.*;" +
                "var input = loadImage(\"images/Autosave001.tif\");" +
                "var gradient = gaussianBlur(input, 200);" +
                "var output = subtractLinear(input, gradient);" +
                "saveImage(output, \"images/Test.png\");"
        engine.eval(script)
    }

    @Throws(IOException::class)
    fun loadImage(filename: String): DoubleImage {
        return ImageReader.read(File(filename), ImageQuality.High)
    }

    @Throws(IOException::class)
    fun saveImage(image: DoubleImage, filename: String?) {
        ImageWriter.write(image, File(filename))
    }

    fun gaussianBlur(image: DoubleImage, radius: Int): DoubleImage {
        return GaussianBlurFilter(radius, ColorModel.RGB).filter(image)
    }

    fun medianFilter(image: DoubleImage, radius: Int): DoubleImage {
        return MedianFilter(radius).filter(image)
    }

    fun horizontalMedianFilter(image: DoubleImage, radius: Int): DoubleImage {
        return HorizontalMedianFilter(radius, ColorModel.RGB).filter(image)
    }

    fun verticalMedianFilter(image: DoubleImage, radius: Int): DoubleImage {
        return VerticalMedianFilter(radius, ColorModel.RGB).filter(image)
    }

    fun pseudoMedianFilter(image: DoubleImage, radius: Int): DoubleImage {
        return PseudoMedianFilter(radius, ColorModel.RGB).filter(image)
    }

    fun gradient(image: DoubleImage): DoubleImage {
        val gradientInterpolationFilter = GradientInterpolationFilter()
        autoSetFixPoints(gradientInterpolationFilter, image)
        return gradientInterpolationFilter.filter(image)
    }

    fun subtractLinear(image1: DoubleImage, image2: DoubleImage): DoubleImage {
        return SubtractLinearImageOperation().operation(image1, image2)
    }
}